Implement the getBlocksFrom() method

AndrejMitrovic opened the pull request:
These two will be used by the block catchup code.

Requires https://github.com/bpfkorea/agora/pull/67

AndrejMitrovic committed:
Add getBlocksFrom() API method

This API method returns the blocks starting from
the provided height, up to the provided number of blocks.

There is a limit of up to MaxBatchBlocksSent of blocks
to be sent with each call to getBlocksFrom();

It's to be used in the catchup phase when a Node must
retrieve blocks it doesn't have in its ledger yet.

AndrejMitrovic committed:
Ledger: Implement getBlocksFrom() method

This method returns the blocks starting from
the provided block height.

The method returns up to 'max_blocks' amount
of blocks.

AndrejMitrovic commented:
The Mac build timed out for some reason. Restarting..

AndrejMitrovic commented:
Mac + ldc 1.16.0

AndrejMitrovic commented:
The mac build is green yet github status still shows it as yellow..

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -13,11 +13,46 @@
+public struct BlockRange
+{
+    ///
+    private Block[Hash] ledger;
```
I think we should move storage to be `Block*[Hash]` and backed by a contiguous list of items, possibly a memory mapped file. Because accessing by height seems to be a very, very common use case, while hash is less common.

AndrejMitrovic replied:
Did you mean `Block*[ulong]` then if you meant accessing by height?

AndrejMitrovic replied:
Well I suppose we should have both Hash and height lookups..

Geod24 replied:
Yeah, accessing by height with an AA lookup could be useful as well. The main point is that it needs to be backed by a contiguous storage, otherwise that's never going to scale. Unless you have other suggestions ?

AndrejMitrovic replied:
I'll think up of something.

AndrejMitrovic commented:
Ok for starters I will split up the getBlockHeight() into another PR so it can be unblocked while the rest of the work is being done on the backing store of the ledger.

Geod24 commented:
So unblocked?

AndrejMitrovic commented:
What I meant was, to unblock getBlockHeight() from having to wait for this PR. Because this PR needs more work.

AndrejMitrovic committed:
Ledger: use an array as a backing store

This increases locality of reference and
makes block lookups and retrieval of chunks
of blocks simple and efficient.

AndrejMitrovic commented:
Fixed the PR. Now it's very simple, and we don't need the hash => block lookup yet, that can be part of a PR that needs it.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -31,6 +33,7 @@ public class Ledger
     /// Ctor
     public this ()
     {
+        this.ledger.reserve(2 ^^ 20);  // reserve ~1 million blocks
```
Yeeeeeaaaaaaaaaahhhhhhhhhhhh... About that... I have mixed feelings about allocating 151Mbs like this with the GC, even if it's just a reserve.

Geod24 commented:
> Fixed the PR. Now it's very simple, and we don't need the hash => block lookup yet, that can be part of a PR that needs it.

It would be less than 10 LoCs to do it in this PR, right ?

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -65,6 +71,30 @@ public class Ledger
+
+    ***************************************************************************/
+
+    public Block[] getBlocksFrom ( ulong block_height, size_t max_blocks )
```
```suggestion
    public Block[] getBlocksFrom (ulong block_height, size_t max_blocks)
```

Geod24 replied:
Actually:

```suggestion
    public Block[] getBlocks ( ulong from, size_t max_blocks )
```

Wouldn't this be better ?

AndrejMitrovic commented:
I've removed the reserve call.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -65,6 +70,30 @@ public class Ledger
+    {
+        assert(max_blocks > 0);
+
+        if (block_height > this.ledger.length)
```
It should be specified in the doc that it's an inclusive range.

AndrejMitrovic replied:
Inclusive in what regards? It starts at the block height and goes from there. And the ledger always has at least one block (genesis).

Geod24 replied:
That the provided `block_height` is included.
So if I want to catch-up, I should send `my_block_height + 1`

AndrejMitrovic replied:
I mean it literally says that in the docs. `Get the array of blocks starting from the provided block height.` I guess "starting from" is not descriptive enough. I'll add another note.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -99,3 +129,71 @@ unittest
+        foreach (_; 0 .. count)
+        {
+            auto rand_pair = KeyPair.random();
+            auto input = Input(hashFull(42));
```
Use the predicted height here ?

AndrejMitrovic replied:
> Use the predicted height

Sorry what?

Geod24 replied:
Well it's obv... Nvm
I originally wrote "use the index" to have a bit more randomness.
Then realized that if you call this function twice, you'll have twice the same set of txs.
Hence, use the height at which this tx is supposed to be included.

AndrejMitrovic replied:
> I originally wrote "use the index" to have a bit more randomness.

"Every time I hear about randomness in blockchain.."

Geod24 reviewed (changes requested):
Can you add a Network Test ?

AndrejMitrovic commented:
> Can you add a Network Test ?

Oh good point!

AndrejMitrovic commented:
> Can you add a Network Test ?

Btw should I add a `Ledger` module in /test? I'd hate to jam even more test code into GossipProtocol.d or elsewhere.

Geod24 commented:
Go with the way you feel most comfortable with. I expect us to refactor / consolidate tests at some point.

AndrejMitrovic commented:
Ah wait.. I didn't add the API part, it got lost in another branch lol.

AndrejMitrovic commented:
Updated. @Geod24 I added a limit to the number of blocks that can be sent at once. I think it's useful to have, and also it should probably be configurable later.

Geod24 commented on source/agora/common/API.d:
```diff
@@ -136,4 +140,26 @@ public interface API
+
+    ***************************************************************************/
+
+    public Block[] getBlocksFrom ( ulong block_height, size_t max_blocks );
```
Are you for real?

Geod24 commented on source/agora/common/API.d:
```diff
@@ -22,6 +23,9 @@ import vibe.data.json;
 import vibe.http.common;
 
+/// Maximum number of blocks that can be sent in a call to getBlocksFrom()
+enum MaxBatchBlocksSent = 1000;
```
```suggestion
private enum MaxBatchBlocksSent = 1000;
```

But it's an implementation details so it should go in `Node` IMO

AndrejMitrovic replied:
Well not really, because clients have to know what the limit is when they request blocks. Otherwise they'll get errors.

Geod24 replied:
Just make it return less instead of erroring.

Geod24 commented on source/agora/common/API.d:
```diff
@@ -136,4 +140,26 @@ public interface API
+    /***************************************************************************
+
+        Get the array of blocks starting from the provided block height.
+        The block at `block_height` is included in the array.
```
```suggestion

        The block at `block_height` is included in the array.
        Note that an implementation is free to return less blocks than asked for.
        It is however required to never return more blocks than asked for.        
```

AndrejMitrovic commented:
Ok fixed the styling.

Geod24 commented on source/agora/test/Ledger.d:
```diff
@@ -0,0 +1,72 @@
+import agora.test.Base;
+
+///
+unittest
```
This is beautiful

AndrejMitrovic replied:
_suspicious_

Geod24 commented on source/agora/test/Ledger.d:
```diff
@@ -0,0 +1,72 @@
+    assert(blocks.length == 1 && blocks[0] == getGenesisBlock());
+
+    blocks = node_1.getBlocksFrom(100, 1);
+    assert(blocks.length == 1 && blocks[0].tx == txes[$ - 1]);
```
Try to get more blocks than there is
Try to start > the max height

Geod24 replied:
That comment was in the network part :'(

AndrejMitrovic replied:
Huh?

Geod24 replied:
You added the tests to the `unittests`, not the network UT.

AndrejMitrovic replied:
See this is why I dislike having modules with the same name..

AndrejMitrovic commented:
Thanks, updated.

AndrejMitrovic commented:
Added the tests to the network unittest.

AndrejMitrovic commented:
@Geod24 anything else?
